pkg/resource/generator.go

```diff
@@ -86,8 +96,132 @@ func (g *Generator) list(cr *regopapi.ImageRegistry) ([]Mutator, error) {
 	return mutators, nil
 }
 
+// syncStorage checks:
+// 1.)  to make sure that an existing storage medium still exists and we can access it
+// 2.)  to see if the storage medium name changed and we need to:
+//      a.) check to make sure that we can access the storage or
+//      b.) see if we need to try to create the new storage
+func (g *Generator) syncStorage(cr *regopapi.ImageRegistry, modified *bool) error {
+
+	// If the Image Registry is currently using S3 storage
+	if cr.Spec.Storage.S3 != nil {
+		// Create a new driver with the current configuration
+		// This saves us from having to duplicate code
+		driver := storageS3.NewDriver(cr.Name, cr.Namespace, cr.Spec.Storage.S3)
+
+		// If the bucket name has changed
+		if cr.Status.Storage.State.S3 != nil && cr.Spec.Storage.S3.Bucket != cr.Status.Storage.State.S3.Bucket {
+			driver.Config.Bucket = cr.Spec.Storage.S3.Bucket
+			// Check to see if the bucket exists
+			// and if we can access it
+			if err := driver.CheckBucketExists(cr); err != nil {
+				installConfig, err := clusterconfig.GetInstallConfig()
+				if err != nil {
+					return err
+				}
+				// If the bucket doesn't exist, try to create it
+				if err := driver.CreateAndTagBucket(installConfig, cr); err != nil {
+					return err
+				}
+				// If we got here, image registry resource
+				// was modified by CreateAndTagBucket
+				*modified = true
+			}
+		} else if cr.Spec.Storage.S3.Bucket != "" {
+			// If the bucket name didn't change,
+			// check to see if the current bucket exists
+			if err := driver.CheckBucketExists(cr); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// syncSecrets checks to see if we have updated storage credentials from:
+// 1.) cluster wide cloud credentials
+// 2.) user provided credentials in the image-registry-private-configuration-user secret
+// and updates the image-registry-private-configuration secret that provides
+// those credentials to the registry pod
+func (g *Generator) syncSecrets(cr *regopapi.ImageRegistry, modified *bool) error {
+	client, err := clusterconfig.GetCoreClient()
+	if err != nil {
+		return err
+	}
+
+	appsclient, err := kappsset.NewForConfig(g.kubeconfig)
+	if err != nil {
+		return err
+	}
+
+	cfg, err := clusterconfig.GetAWSConfig()
+	if err != nil {
+		return err
+	}
+
+	// Get the existing image-registry-private-configuration secret
+	sec, err := client.Secrets(g.params.Deployment.Namespace).Get(regopapi.ImageRegistryPrivateConfiguration, metav1.GetOptions{})
```

Comment:
as a rule for content we are already watching, and thus have cached in a listwatcher, we should retrieve it from the cache, not make a new api call.  There are a few spots in this PR that this applies to.  I'm not going to make you fix it in this PR, but it's something to keep in mind and come back to.

Reply:
ok, i will keep that in mind in the future